### DISCUSSION TITLE
Comment comparing dosemu2 KVM/V86 to qemu KVM/R86, and generally

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#2064] [General]
I replied to a user on the FreeDOS sourceforge.net feature tracker today. The following is my text in full. Do you have anything to add or correct about the bits and bobs I wrote there? Feel free to reply here or over there.

----

> If you should decide for a special build I would suggest ldebugr as a name, the r then stands for real mode without DPMI. As I assume you want to omit the PM or DPMI part in this particular build to keep the program under 64kb.

lDebug without the X already is the 86-Mode-only variant. So if I want to add a more feature-rich build it should have a name indicating that it has more options enabled, not simply a trait that it shares with the non-X default lDebug build. (As an aside, the script names like `makerd` already use the R letter naming for the non-X builds, just not in the build name they eventually create.)

> On the other hand, if only the PM/DPMI variant LDEBUX is affected, couldn't one simply swap code parts of the LDEBUX variant into the XMS memory?

(I will reply assuming a DPMI memory allocation, which is less of a headache to access while in Protected Mode than an XMS allocation.) This would be no easier than the dual code segments approach that we already have, and would fail more often, as in you couldn't access the DPMI allocated code segment while not yet in Protected Mode, or while switched temporarily to 86 Mode.

In case you weren't aware, the dual code segments option splits off some code parts into a second code segment. This is possible in all builds of lDebug; in 86-Mode-only builds the debugger relocates a bunch of far calls (similar to MZ .EXE file relocations) during installation (from the init section), while in Protected-Mode-enabled builds all calls to dual-distance functions are done with calls to helpers that resolve the far target and return addresses in agreement with the current mode (segments in 86 Mode, selectors in Protected Mode).

So we do have a way to lower the code segment pressure, it is simply a hassle both in the work to write and use it and in the space and performance hits associated with the dual calls.

> Since dosemu2 allows you to copy and paste the text from the DOS environment into the Linux host system, I would even assume that DOSEMU2 is more of a kind of a wrapper than a full emulator.
> If the DOS application then tries to write into foreign memory areas, I wouldn't be surprised if the Linux kernel then responds with a sigkill. But that's just a guess on my part.

This is half accurate, but only just so.

For 86 Mode, both qemu and dosemu2 can use KVM, but qemu will run the client in Real 86 Mode and pass along all exceptions to the client, it appears. For dosemu2, the client still runs in Virtual 86 Mode, and some exceptions will terminate dosemu rather than be passed to the client. (lDebug's RV command can display one of four modes: Real 86 Mode, Virtual 86 Mode, or DPMI with a 16- or 32-bit code selector.)

For example, running an instruction like `mov ax, word [FFFF]` in qemu KVM will cause a Real 86 Mode General Protection Fault (interrupt 0Dh) to the client.

(An lDebug build with the build option `-D_CATCHINT0D` can be used to intercept this; without such a special handler the IRQ #5 hardware interrupt handler will be repeatedly called, easily resulting in an infinite loop. Another option to get out of the infinite loop is to `install timer, serial` and break back into the debugger using a double Control-C to the serial input. Yet another option is to build with `-D_CATCHSYSREQ -D_SYSREQINT=15h` and try to break to the debugger using the SysReq key. However, in qemu version `QEMU emulator version 7.2.2 (Debian 1:7.2+dfsg-7)` this seems to lock the keyboard input from reaching the debugger. Instead, we can try to run like (faulting instruction at 60h:0) `rc.replace g; tp 100 while csip != 60_0000 silent 1` then `rc`. This last one doesn't work properly in qemu either; even with the work-around attempt (the TP command running from RC buffer) it crashes the debugger and/or debuggee somewhat, probably due to inreentrant IRQ handlers in their ROM-BIOS. I had a similar problem on dosemu2 (the first problem, int 15h SysReq detection locking the keyboard input IRQ), which [got fixed on their end](https://github.com/dosemu2/dosemu2/issues/1780) by modifying their ROM-BIOS handler for interrupt 09h (IRQ #1). Using `-D_SYSREQINT=09h` never seems to detect the Alt+Druck (S-Abf, ie SysReq) keypress in this version of qemu.)

(In non-KVM qemu, segment limits are never checked.)

In dosemu2 KVM, the same faulting instruction will crash dosemu2 because the VMM doesn't handle it the same way.

This is not really a difference of how much of a "wrapper" or "full emulator" either program is, the faults are simply handled differently.

I do use "native DPMI" mode for dosemu2, which is probably faster than any alternative. To my understanding, it runs your DPMI Protected Mode code basically as native Linux code, with the descriptors set up to allow 16-bit and 32-bit Protected Mode execution (which still both work even with the OS in amd64 Long Mode). I'm unsure about how much this changes the big picture on faults, as some can still clearly be caught by dosemu2 and reflected into the client. (For instance, when I run `lcdebugx.com /d- lcdebugx.com dpmimini.com` then enter PM, the inner lCDebugX may cause a Page Fault accessing invalid memory at >= 1088 KiB and this will be caught by the outer lCDebugX.)

But this mode is certainly closer to regular Linux host user mode than whatever qemu does, KVM or no. On the other hand, this is not a mode used by default, you have to enable it explicitly, eg using the dosemu2 command line switch `-I "cpu_vm_dpmi native"`.

**Comment by ecm-pushbx** [comment]
Link would be helpful, I forgot to add it: https://sourceforge.net/p/freedos/feature-requests/118/?page=1#05c4/e111/feb5/52b0/0211/3f54/4b82/03df/22a3/fa2a/cd75/b00c/3121/8e3b/0525/6e08/4adb/4cab

**Comment by stsp** [comment]
> (I will reply assuming a DPMI memory allocation, which is less of a headache to access while in
> Protected Mode than an XMS allocation.)

Under dosemu2 the XMS calls are
translated, and in a very convenient
way. See our `test/func_memory_xms.py`
as an example, but I don't know if anyone
else is doing such translation. We do more
translations than even HX does.

> would even assume that DOSEMU2 is more of a kind of a wrapper than a full emulator.

This is something I really have in mind.
I want to leverage the winelib-alike approach
eventually, and lots of work in that direction
is already done, but its not happening quickly.
So for now dosemu2 is a "kinda" full emulator. :)
But there is the other side, like not running the
ring-0 code and instead executing the DPMI
client directly. Like mapping in the host FS
via redirector, like providing the int33 mouse
driver bypassing the mouse HW emulation.
Like emulating memory subsystems (XMS, EMS)
directly on top of host's VMM.
Many devices are "weirdly" emulated, i.e.
instead of emulating the HW timers as they
are, we have virtual timer devices + bios
code + host code, which is visible from the
DOS side mostly as a legacy timers, unless
you know the extra IO ports to access the
underlying devs instead. I constantly request
or provide the various extensions to linux
kernel or glibc to better map the DOS
subsystems directly to host (from recent
examples that would be OFD locking
extension proposal to kernel and posix
timers proposal to glibc).
And so on, so forth.
So I am really pushing dosemu2 towards
the "thin wrapper" approach, and obviously
fdpp was the major sign that we are going
to follow dosbox, not dosemu1 or qemu.

> In dosemu2 KVM, the same faulting instruction will crash dosemu2 because the VMM doesn't
> handle it the same way.

Not correct: you can use DPMI fn 0x213
if you want the real-mode faults to go to
the client. Although I think 0x213 didn't
work correctly until... today. Hooking an
RM ints via IVT should also work for the
first 8 exceptions (from 0 to 7).

> I do use "native DPMI" mode for dosemu2, which is probably faster than any alternative.

No, its not faster than KVM, because in KVM
you can use native IDT and get fewer traps.
Also native dpmi switch involves signal and
sigreturn(), which are slow. Also native dpmi
is insecure, and also you can't hook the
interrupts that are used by the host kernel
(like 0x80) because we can't change native
IDT (but we can change KVM's idt).
If we make native dpmi secure, it would
be even slower than now. So there is no
reason to use it.

**Reply by ecm-pushbx** [reply]
Hmm, but native DPMI mode is the fastest if KVM is not available to dosemu2, right?

**Reply by ecm-pushbx** [reply]
> Not correct: you can use DPMI fn 0x213
if you want the real-mode faults to go to
the client. Although I think 0x213 didn't
work correctly until... today. Hooking an
RM ints via IVT should also work for the
first 8 exceptions (from 0 to 7).

But in qemu's (KVM) Real 86 Mode I can catch the GPF or stack fault using the special int 0Dh / int 0Ch handlers, which check whether they are called as IRQ #5 / IRQ #4, or not (assumed to mean a fault). No need to init or ever switch to PM.

**Reply by ecm-pushbx** [reply]
Here's the code which checks for an IRQ #5 versus the GPF from qemu's Real 86 Mode. (The same method works on actual hardware CPUs too.) https://hg.pushbx.org/ecm/ldebug/file/eab8e19cb957/source/run.asm#l6147

**Reply by stsp** [reply]
> Hmm, but native DPMI mode is the fastest if KVM is not available to dosemu2, right?

Sure thing.
And that will hopefully not change even
if we sandbox it properly, but it will certainly
became slower than now.

> But in qemu's (KVM) Real 86 Mode I can catch the GPF or stack fault using the special int 0Dh / int 0Ch
> handlers

I am not sure why its not the case here.
You can open a bug for that, or better yet,
a PR that changes vm86_fault() by adding
more exceptions to the initial `switch` block.

**Reply by ecm-pushbx** [reply]
I tried that, but it isn't enough:

```
dosemu2$ git diff
diff --git a/src/base/emu-i386/do_vm86.c b/src/base/emu-i386/do_vm86.c
index a20c5b452..f21302ccd 100644
--- a/src/base/emu-i386/do_vm86.c
+++ b/src/base/emu-i386/do_vm86.c
@@ -70,6 +70,8 @@ int vm86_fault(unsigned trapno, unsigned err, dosaddr_t cr2)
   case 0x04: /* overflow */
   case 0x05: /* bounds */
   case 0x07: /* device_not_available */
+  case 0x0C:
+  case 0x0D:
     do_int(trapno);
     return 0;
 
```

Still same fault exit from dosemu2 (KVM):

```
Welcome to dosemu2!
    Build 2.0pre9-dev-20230805-1380-gf8e24bdab
lDebug (2023-08-05)
-r
AX=0000 BX=0000 CX=000B DX=0000 SP=FFFE BP=0000 SI=0000 DI=0000
DS=2E89 ES=2E89 SS=2E89 CS=2E89 IP=0100 NV UP EI PL ZR NA PE NC
2E89:0100 A1FFFF            mov     ax, [FFFF]                     DS:FFFF=CD00
-t
Security warning: native DPMI mode is insecure, adjust $_cpu_vm_dpmi
ERROR: MIDI: unsupported synth mode mt32
ERROR: KVM: applying TF fixup
ERROR: KVM: applying TF fixup
ERROR: general protection at 0x4202e990: a1
```

**Reply by ecm-pushbx** [reply]
Managed to get it done using the following:

```
dosemu2$ git diff
diff --git a/src/base/emu-i386/do_vm86.c b/src/base/emu-i386/do_vm86.c
index a20c5b452..9493c171e 100644
--- a/src/base/emu-i386/do_vm86.c
+++ b/src/base/emu-i386/do_vm86.c
@@ -70,6 +70,8 @@ int vm86_fault(unsigned trapno, unsigned err, dosaddr_t cr2)
   case 0x04: /* overflow */
   case 0x05: /* bounds */
   case 0x07: /* device_not_available */
+  case 0x0C:
+  case 0x0D:
     do_int(trapno);
     return 0;
 
@@ -421,6 +423,8 @@ static void vm86_GP_fault(void)
 #ifdef USE_MHPDBG
     mhp_debug(DBG_GPF, 0, 0);
 #endif
+    do_int(0x0D);
+    return;
     set_debug_level('g', 1);
     error("general protection at %p: %x\n", lina,*lina);
     show_regs();
```

**Reply by ecm-pushbx** [reply]
Both interrupts 0Ch and 0Dh are generated using this, depending on whether an `ss` prefix is present or not.
